File: OLD Repos/Numerical Methods/Lagrange Interpolation Formula/Lagrange Interpolation Formula.py
#### Lagrange Interpolation Formula

import time
import numpy as np
import matplotlib.pyplot as plt
from scipy.interpolate import lagrange

to = time.time()

def lagrange_interpolate(x, y, t):
    p = lagrange(x, y)
    return p(t)

x, y = np.loadtxt('Data/dataSim.dat', unpack=True)
t = np.linspace(1, 2, 11)
u = lagrange_interpolate(x, y, t)

td = time.time() - to
print("The time interval is %f s." %td)

plt.rcParams['font.family'] = 'CMU Serif'
plt.plot(t, u, "P", c="#00B4DC", alpha=0.9)
plt.plot(x, y, "o", c="#32B432", alpha=0.9)
plt.tick_params(direction='in')
plt.show()

File: OLD Repos/Numerical Methods/Lagrange Interpolation Formula/Lagrange Interpolation Formula.jl
#### Lagrange Interpolation Formula

using DelimitedFiles
using Plots; pyplot()

to = time()

function lagrange_interpolate(X,Y,t)
    C = ones(length(X))
    d = 0.0
    for i = 1:length(X)
        for j = [1:i-1;i+1:length(X)]
            C[i] = C[i]*(t-X[j])/(X[i]-X[j])
        end
        d = d + Y[i]*C[i]
    end
    return d
end

function lagrange_interpolate_plus(X,Y,t)
    idxs = eachindex(X)
    sum(Y[i] * prod((t-X[j])/(X[i]-X[j]) for j in idxs if j != i) for i in idxs)
end

A = readdlm("Data/dataSim.dat")
X = view(A,:,1)
Y = view(A,:,2)
T = 1.0:0.1:2.0
U = [lagrange_interpolate_plus(X, Y, t) for t in T]

td = time() - to
println("The time interval is $td s.")

plot(fontfamily=("Serif"),dpi=512)
# scatter!(X,Y,color="#32B432",marker=(:circle,10,Plots.stroke(:white)),label="Source")
# scatter!(T,U,color="#00B4DC",marker=(:cross,10,Plots.stroke(:white)),label="Foreast")
plot!(X,Y,color="#32B432",marker=(:circle,10,Plots.stroke(:white)),label="Source")
plot!(T,U,color="#00B4DC",marker=(:cross,10,Plots.stroke(:white)),label="Foreast")

# savefig("U.pdf")